File: aiweb/src/main/js/geneticprogramming/TreeVisitor.js
define(function()
{
    "use strict";
    function TreeVisitor(treeNode)
    {
        InputValidator.validateNotNull("treeNode", treeNode);

        var level = 0;
        var nodeCount = 0;
        var sb = "";

        this.description = function()
        {
            return sb;
        };

        this.nodeCount = function()
        {
            return nodeCount;
        };

        this.visit = function(treeNode)
        {
            for (var i = 0; i < level; i++)
            {
                sb += "  ";
            }
            sb += nodeCount;
            sb += ": ";
            sb += treeNode.symbol();
            sb += " ";
            sb += treeNode.name();
            sb += "\n";

            nodeCount++;

            if (treeNode.arity() > 0)
            {
                level++;
                var size = treeNode.arity();

                for (var j = 0; j < size; j++)
                {
                    var child = treeNode.childAt(j);
                    child.accept(this);
                }

                level--;
            }
        };

        this.visit(treeNode);
    }

    return TreeVisitor;
});

File: ai/src/main/java/org/vizzini/ai/geneticalgorithm/geneticprogramming/TreeVisitor.java
package org.vizzini.ai.geneticalgorithm.geneticprogramming;

import org.apache.commons.lang3.StringUtils;
import org.vizzini.core.Visitor;

/**
 * Provides a visitor to print a tree as a tree.
 * 
 * @param <T> Type.
 */
public final class TreeVisitor<T> implements Visitor<TreeNode<T>>
{
    /**
     * @param treeNode Tree node.
     * 
     * @return the description.
     */
    public static <T> String toDescription(final TreeNode<T> treeNode)
    {
        final TreeVisitor<T> visitor = new TreeVisitor<T>();
        treeNode.accept(visitor);

        return visitor.getDescription();
    }

    /** Level. */
    private int level = 0;

    /** Node count. */
    private int nodeCount = 0;

    /** String builder. */
    private StringBuilder sb = new StringBuilder();

    /**
     * @return the description.
     */
    public String getDescription()
    {
        return sb.toString();
    }

    /**
     * @return the nodeCount
     */
    public int getNodeCount()
    {
        return nodeCount;
    }

    @Override
    public void visit(final TreeNode<T> treeNode)
    {
        sb.append(StringUtils.repeat("  ", level));
        sb.append(nodeCount);
        sb.append(": ");
        sb.append(treeNode.getSymbol());
        sb.append(" ");
        sb.append(treeNode.getClass().getSimpleName());
        sb.append("\n");

        nodeCount++;

        if (treeNode instanceof Function)
        {
            level++;

            final Function<T> function = (Function<T>)treeNode;
            final int size = function.getArity();

            for (int i = 0; i < size; i++)
            {
                final TreeNode<T> child = function.getChildAt(i);
                child.accept(this);
            }

            level--;
        }
    }
}
